jwt-angular/src/app/auth.service.ts

```diff
@@ -0,0 +1,135 @@
+import { Injectable } from '@angular/core';
+import { Observable, of } from 'rxjs';
+import { tap, delay } from 'rxjs/operators';
+// esModuleInterop in your tsconfig
+import moment from 'moment';
+
+import { ApiService } from './api.service';
+
+export interface AuthData {
+  username: string;
+  password: string;
+}
+
+export interface AuthResponse {
+  access: string;
+  refresh?: string;
+  access_expiry: number;
+  refresh_expiry?: number;
+  csrf_token: string;
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class AuthService {
+  // Local storage
+  private ls = window.localStorage;
+  // Store auth info in memory
+  // CSRF token
+  csrfToken: string;
+  // Access expiry
+  private accessExpiry: number;
+  // Refresh expiry
+  private refreshExpiry: number;
+
+  constructor(private apiService: ApiService) {
+    // Syncronize other tabs
+    window.addEventListener('storage', this.syncAuth);
+  }
+
+  // Sign in
+  obtainBoth(data: AuthData): Observable<AuthResponse> {
+    return this.apiService.post(`/token/both/`, data)
+      .pipe(tap(response => {
+        this.setAuth(response);
+        this.scheduleRefresh();
```

Comment:
So I have implemented a subscription cancellation via `cancelRefresh` which is called in `purgeAuth`.

You mention signing the user out at the second 401 request you would get in a row. I think this makes sense and I can bake in something like this.

Reply:
Yes, that'd be helpful.

Just to clarify, this is because JWT intends to deter those who capture a token in an insecure network (i.e. server using HTTP instead of HTTPS).

Reply:
Yes 2 reasons, avoid generating unnecessary traffic to your back-end and the one you mention above (which I didn't think about to be honest :smile_cat: )